File: helper/Sources/ShootrKit/Sharpness.swift
import Foundation

/// Tenengrad (Sobel gradient energy) over Laplacian variance: less
/// noise-sensitive, which matters at the ISO 3200–12800 that indoor wedding
/// work lives at (design 03 §3.2). Pure math — fully unit-testable.
public enum Sharpness {

    /// Mean squared Sobel gradient magnitude over the buffer, normalized so
    /// values are comparable across crops. Absolute numbers are meaningless;
    /// only ratios are diagnostic (design 03 §3.1).
    public static func tenengrad(
        pixels: [UInt8], width: Int, height: Int
    ) -> Double {
        guard width > 2, height > 2 else { return 0 }
        var sum = 0.0
        for y in 1..<(height - 1) {
            let up = (y - 1) * width, mid = y * width, dn = (y + 1) * width
            for x in 1..<(width - 1) {
                let a = Double(pixels[up + x - 1]), b = Double(pixels[up + x])
                let c = Double(pixels[up + x + 1])
                let d = Double(pixels[mid + x - 1]), f = Double(pixels[mid + x + 1])
                let g = Double(pixels[dn + x - 1]), h = Double(pixels[dn + x])
                let i = Double(pixels[dn + x + 1])
                let gx = (c + 2 * f + i) - (a + 2 * d + g)
                let gy = (g + 2 * h + i) - (a + 2 * b + c)
                sum += gx * gx + gy * gy
            }
        }
        let n = Double((width - 2) * (height - 2))
        // Max |g| per axis is 4*255; normalize energy into a sane 0..1-ish range.
        return sum / n / (4.0 * 255 * 255)
    }

    public struct TileMap: Codable {
        public let tiles: [[Double]]  // 16×16, row-major
        public let max: Double
        public let mean: Double
    }

    /// 16×16 tile grid (design 03 §3.3): normalization denominator,
    /// landscape focus-plane detection, motion-blur detection.
    public static func tileMap(
        pixels: [UInt8], width: Int, height: Int, grid: Int = 16
    ) -> TileMap {
        var tiles = [[Double]](
            repeating: [Double](repeating: 0, count: grid), count: grid)
        var maxV = 0.0, sum = 0.0
        let tw = width / grid, th = height / grid
        guard tw > 2, th > 2 else {
            return TileMap(tiles: tiles, max: 0, mean: 0)
        }
        for ty in 0..<grid {
            for tx in 0..<grid {
                var tile = [UInt8](repeating: 0, count: tw * th)
                let ox = tx * tw, oy = ty * th
                for row in 0..<th {
                    let src = (oy + row) * width + ox
                    tile.replaceSubrange(
                        row * tw..<(row + 1) * tw,
                        with: pixels[src..<src + tw])
                }
                let v = tenengrad(pixels: tile, width: tw, height: th)
                tiles[ty][tx] = v
                maxV = Swift.max(maxV, v)
                sum += v
            }
        }
        return TileMap(
            tiles: tiles, max: maxV, mean: sum / Double(grid * grid))
    }
}

File: analyzer/shootr_analyzer/sharpness.py
"""Tenengrad sharpness — numeric port of helper/Sources/ShootrKit/Sharpness.swift.

The two analyzers must agree on this math or the A/B comparison measures the
port, not the decode. Same Sobel kernels, same interior-only iteration, same
normalization sum/n/(4·255)² — validated by running the Swift selftest's
synthetic patterns through both (tests/test_sharpness.py, selftest.py).

Vectorized with numpy but kept semantically identical: mean squared gradient
magnitude over the (w-2)×(h-2) interior.
"""

from __future__ import annotations

from dataclasses import dataclass

import numpy as np

TILE_GRID = 16


def tenengrad(pixels: np.ndarray) -> float:
    """Mean squared Sobel gradient magnitude, normalized to a 0..1-ish range.
    `pixels` is a 2-D uint8 array. Absolute numbers are meaningless; only
    ratios are diagnostic (design 03 §3.1)."""
    h, w = pixels.shape
    if w <= 2 or h <= 2:
        return 0.0
    p = pixels.astype(np.float64)
    # 3×3 Sobel via shifted views over the interior — matches the Swift
    # loop's indexing exactly (no border padding, no cv2 border modes).
    a = p[:-2, :-2]; b = p[:-2, 1:-1]; c = p[:-2, 2:]
    d = p[1:-1, :-2];                  f = p[1:-1, 2:]
    g = p[2:, :-2];  hh = p[2:, 1:-1]; i = p[2:, 2:]
    gx = (c + 2 * f + i) - (a + 2 * d + g)
    gy = (g + 2 * hh + i) - (a + 2 * b + c)
    total = float(np.sum(gx * gx + gy * gy))
    n = float((w - 2) * (h - 2))
    return total / n / (4.0 * 255 * 255)


@dataclass
class TileMap:
    tiles: list[list[float]]  # 16×16, row-major
    max: float
    mean: float


def tile_map(pixels: np.ndarray, grid: int = TILE_GRID) -> TileMap:
    """16×16 tile grid (design 03 §3.3): normalization denominator, landscape
    focus-plane detection, motion-blur detection. Tiles are cropped views of
    size (h//grid, w//grid), same truncation as the Swift version — trailing
    rows/columns beyond grid*tile are ignored by both."""
    h, w = pixels.shape
    tw, th = w // grid, h // grid
    if tw <= 2 or th <= 2:
        zero = [[0.0] * grid for _ in range(grid)]
        return TileMap(tiles=zero, max=0.0, mean=0.0)
    tiles = [[0.0] * grid for _ in range(grid)]
    max_v = 0.0
    total = 0.0
    for ty in range(grid):
        for tx in range(grid):
            v = tenengrad(pixels[ty * th:(ty + 1) * th, tx * tw:(tx + 1) * tw])
            tiles[ty][tx] = v
            max_v = max(max_v, v)
            total += v
    return TileMap(tiles=tiles, max=max_v, mean=total / (grid * grid))


def clipping(pixels: np.ndarray) -> tuple[float, float]:
    """Fraction of pixels at the rails — same thresholds as Analyze.swift
    (>=254 high, <=1 low)."""
    n = float(pixels.size)
    hi = float(np.count_nonzero(pixels >= 254)) / n
    lo = float(np.count_nonzero(pixels <= 1)) / n
    return hi, lo
